src/optlang/matrix_interface.py

```diff
@@ -0,0 +1,961 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generic solver interface for (sparse) matrix solvers.
+
+This interface is only supposed to be a basis to add matrix solvers to optlang. It
+creates a thin layer between all optlang operations and the standard formulation of
+LP, QP, and MILP problems.
+
+To create a solver from this, you need to do the following:
+
+1. Create a class that inherits from MatrixProblem and implement the `solve` method
+   and optionally overwrite the `reset` and `still_valid` methods.
+
+2. Create derived classes for Variable, Constraint, Objective, and Configuration.
+   You don't need to overwrite anything. Those should work out of the box.
+
+3. Create a derived class for Model and set the `ProblemClass` attribute to the class
+   from (1) and the `status_map` to map the solver status to an Optlang status from
+   `optlang.interface`.
+
+For an example take a look at `optlang.osqp_interface`.
+"""
+
+import logging
+import six
+import pickle
+from collections import defaultdict
+from numpy import array, concatenate, Infinity, zeros
+
+from optlang import interface, symbolics, available_solvers
+from optlang.util import inheritdocstring
+from optlang.expression_parsing import parse_optimization_expression
+from optlang.exceptions import SolverError
+
+from scipy.sparse import csc_matrix
+
+from optlang.symbolics import add, mul
+
+log = logging.getLogger(__name__)
+
+
+_STATUS_MAP = defaultdict(lambda: interface.UNDEFINED)
+
+_LP_METHODS = ("auto", )
+
+_QP_METHODS = ("auto", )
+
+_TYPES = ("continuous", "binary", "integer")
+
+
+class MatrixProblem(object):
```

Comment:
Would it make sense to mark this as an abstract class?

```suggestion
class MatrixProblem(abc.ABC):
```

Reply:
Done.